File: 2015/05/rust/src/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants ---
const INPUT_FILE:&'static str = "../input/input.txt";
const LIST_1: [char; 5] = ['a', 'e', 'i', 'o', 'u'];
const LIST_2: [&'static str; 4] = ["ab", "cd", "pq", "xy"];

// --- Rules ---
fn rule_1_1(s: &str) -> bool {
    let mut count = 0;
    for c in s.chars() {
       if LIST_1.contains(&c) { count += 1; }
       if count == 3 { return true; }
    }
    return false;
}

fn rule_1_2(s: &str) -> bool {
    let a:Vec<_> = s.chars().collect();
    for i in 0 .. a.len() - 1 {
        if a[i] == a[i + 1] { return true; }
    }
    return false;
}

fn rule_1_3(s: &str) -> bool {
    for c in LIST_2 {
        if s.contains(c) { return false; }
    }
    return true;
}

fn rule_2_1(s: &str) -> bool {
    let a:Vec<_> = s.chars().collect();
    for i in 0 .. a.len() - 2 {
        for j in i + 2 .. a.len() - 1 {
            if a[i] == a[j] && a[i + 1] == a[j + 1] { return true; }
        }
    }
    return false;
}

fn rule_2_2(s: &str) -> bool {
    let a:Vec<_> = s.chars().collect();
    for i in 0 .. a.len() - 2 {
        if a[i] == a[i + 2] { return true; }
    }
    return false;
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut count1 = 0;
    let mut count2 = 0;

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
        if rule_1_1(&line) && rule_1_2(&line) && rule_1_3(&line) { count1 += 1; }
        if rule_2_1(&line) && rule_2_2(&line) { count2 += 1; }
    }

    // --- Puzzle 1 ---
    println!("1. Nice Strings: {}", count1);

    // --- Puzzle 2 ---
    println!("2. Nice Strings: {}", count2);

    Ok(())
}

File: 2015/05/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"

var LIST1 = []rune{'a', 'e', 'i', 'o', 'u'}
var LIST2 = []string{"ab", "cd", "pq", "xy"}

// --- Rules ---
func rule11(s string) bool {
	count := 0
	for _, c := range s {
		if slices.Contains(LIST1, c) {
			count++
		}
		if count == 3 {
			return true
		}
	}
	return false
}

func rule12(s string) bool {
	a := []rune(s)
	for i := 0; i < len(a)-1; i++ {
		if a[i] == a[i+1] {
			return true
		}
	}
	return false
}

func rule13(s string) bool {
	for _, c := range LIST2 {
		if strings.Contains(s, c) {
			return false
		}
	}
	return true
}

func rule21(s string) bool {
	a := []rune(s)
	for i := 0; i < len(a)-2; i++ {
		for j := i + 2; j < len(a)-1; j++ {
			if a[i] == a[j] && a[i+1] == a[j+1] {
				return true
			}
		}
	}
	return false
}

func rule22(s string) bool {
	a := []rune(s)
	for i := 0; i < len(a)-2; i++ {
		if a[i] == a[i+2] {
			return true
		}
	}
	return false
}

func main() {
	// --- Variables --
	count1 := 0
	count2 := 0

	// --- Read input file ---
	file, err := os.Open(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if rule11(line) && rule12(line) && rule13(line) {
			count1++
		}
		if rule21(line) && rule22(line) {
			count2++
		}
	}

	// --- Results ---
	fmt.Println("1. Nice strings:", count1)
	fmt.Println("2. Nice strings:", count2)
}
